src/app/SubscriptionResumptionSessionEstablisher.cpp

```diff
@@ -0,0 +1,112 @@
+        {
+            mSubscriptionInfo.mEventPaths[i] = subscriptionInfo.mEventPaths[i];
+        }
+    }
+
+    ScopedNodeId peerNode = ScopedNodeId(mSubscriptionInfo.mNodeId, mSubscriptionInfo.mFabricIndex);
+    caseSessionManager.FindOrEstablishSession(peerNode, &mOnConnectedCallback, &mOnConnectionFailureCallback);
+    return CHIP_NO_ERROR;
+}
+
+void SubscriptionResumptionSessionEstablisher::HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                                                     const SessionHandle & sessionHandle)
+{
+    SubscriptionResumptionSessionEstablisher * _this = static_cast<SubscriptionResumptionSessionEstablisher *>(context);
+    SubscriptionResumptionStorage::SubscriptionInfo & subscriptionInfo = _this->mSubscriptionInfo;
+    InteractionModelEngine * imEngine                                  = InteractionModelEngine::GetInstance();
+    if (!imEngine->EnsureResourceForSubscription(subscriptionInfo.mFabricIndex, subscriptionInfo.mAttributePaths.AllocatedSize(),
+                                                 subscriptionInfo.mEventPaths.AllocatedSize()))
+    {
+        ChipLogProgress(InteractionModel, "no resource for subscription resumption");
+        delete _this;
+        return;
+    }
+    ReadHandler * readHandler = imEngine->mReadHandlers.CreateObject(*imEngine, imEngine->GetReportScheduler());
+    if (readHandler == nullptr)
+    {
+        ChipLogProgress(InteractionModel, "no resource for ReadHandler creation");
+        delete _this;
+        return;
+    }
+    readHandler->OnSubscriptionResumed(sessionHandle, *_this);
+    delete _this;
+}
+
+void SubscriptionResumptionSessionEstablisher::HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId,
+                                                                             CHIP_ERROR error)
+{
+    SubscriptionResumptionSessionEstablisher * _this = static_cast<SubscriptionResumptionSessionEstablisher *>(context);
+    SubscriptionResumptionStorage::SubscriptionInfo & subscriptionInfo = _this->mSubscriptionInfo;
+    ChipLogError(DataManagement, "Failed to establish CASE for subscription-resumption with error '%" CHIP_ERROR_FORMAT "'",
+                 error.Format());
+    // If the device fails to establish the session, the subscriber might be offline and its subscription read client will
+    // be deleted when the device reconnect to the subscriber. This subscription will be never used again. So clean up
+    // the persistent subscription information storage.
+    auto * subscriptionResumptionStorage = InteractionModelEngine::GetInstance()->GetSubscriptionResumptionStorage();
+    if (subscriptionResumptionStorage)
+    {
+        subscriptionResumptionStorage->Delete(subscriptionInfo.mNodeId, subscriptionInfo.mFabricIndex,
+                                              subscriptionInfo.mSubscriptionId);
+    }
+    delete _this;
```

Comment:
same here: arm a deletion using RAII at the top of the function. Maybe that could even have an operator-> and then the code looks like:

```cpp
   AutoDeleteEstablisher establisher(context);
   SubscriptionResumptionStorage::SubscriptionInfo & subscriptionInfo = establisher->mSubscriptionInfo;
   ....
```

Reply:
Done